From 01adb60d7af7420e089b8dc6e15973535fd9ab45 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentin.casasnovas@vates.tech>
Date: Thu, 19 Mar 2026 10:54:45 +0100
Subject: [PATCH] !kabi scsi: cxgb4i: add wait_for_completion()

We need to take care of the changed functors prototypes as well as use the
shadow variable API to add the cmpl field.

Fixes: 4c3b23e90307 ("scsi: cxgb4i: add wait_for_completion()")
Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 drivers/scsi/cxgbi/cxgb4i/cxgb4i.c | 26 +++++++++++++++++++++-----
 drivers/scsi/cxgbi/libcxgbi.c      |  9 +++++++--
 drivers/scsi/cxgbi/libcxgbi.h      | 25 ++++++++++++++++++++++++-
 3 files changed, 52 insertions(+), 8 deletions(-)

diff --git a/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c b/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
index 689d6c813a50..3d1bfaa78f8d 100644
--- a/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
+++ b/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
@@ -11,6 +11,7 @@
  *		Rakesh Ranjan (rranjan@chelsio.com)
  */
 
+#include "linux/shadow_var.h"
 #define pr_fmt(fmt) KBUILD_MODNAME ":%s: " fmt, __func__
 
 #include <linux/kernel.h>
@@ -1515,6 +1516,7 @@ static void do_set_tcb_rpl(struct cxgbi_device *cdev, struct sk_buff *skb)
 	struct cxgb4_lld_info *lldi = cxgbi_cdev_priv(cdev);
 	struct tid_info *t = lldi->tids;
 	struct cxgbi_sock *csk;
+	struct completion *cmpl;
 
 	csk = lookup_tid(t, tid);
 	if (!csk) {
@@ -1532,7 +1534,9 @@ static void do_set_tcb_rpl(struct cxgbi_device *cdev, struct sk_buff *skb)
 		csk->err = -EINVAL;
 	}
 
-	complete(&csk->cmpl);
+	cmpl = shadow_var_get(csk, "shadow_cxgbi_sock");
+	if (cmpl)
+		complete(cmpl);
 
 	__kfree_skb(skb);
 }
@@ -1913,10 +1917,16 @@ static int ddp_setup_conn_pgidx(struct cxgbi_sock *csk, unsigned int tid,
 {
 	struct sk_buff *skb;
 	struct cpl_set_tcb_field *req;
+	struct completion *cmpl;
 
 	if (!pg_idx || pg_idx >= DDP_PGIDX_MAX)
 		return 0;
 
+
+	cmpl = shadow_var_get(csk, "shadow_cxgbi_sock");
+	if (!cmpl)
+		return -ENOMEM;
+
 	skb = alloc_wr(sizeof(*req), 0, GFP_KERNEL);
 	if (!skb)
 		return -ENOMEM;
@@ -1934,9 +1944,10 @@ static int ddp_setup_conn_pgidx(struct cxgbi_sock *csk, unsigned int tid,
 	log_debug(1 << CXGBI_DBG_TOE | 1 << CXGBI_DBG_SOCK,
 		"csk 0x%p, tid 0x%x, pg_idx %u.\n", csk, csk->tid, pg_idx);
 
-	reinit_completion(&csk->cmpl);
+	cmpl = shadow_var_get(csk, "shadow_cxgbi_sock");
+	reinit_completion(cmpl);
 	cxgb4_ofld_send(csk->cdev->ports[csk->port_id], skb);
-	wait_for_completion(&csk->cmpl);
+	wait_for_completion(cmpl);
 
 	return csk->err;
 }
@@ -1946,10 +1957,15 @@ static int ddp_setup_conn_digest(struct cxgbi_sock *csk, unsigned int tid,
 {
 	struct sk_buff *skb;
 	struct cpl_set_tcb_field *req;
+	struct completion *cmpl;
 
 	if (!hcrc && !dcrc)
 		return 0;
 
+	cmpl = shadow_var_get(csk, "shadow_cxgbi_sock");
+	if (!cmpl)
+		return -ENOMEM;
+
 	skb = alloc_wr(sizeof(*req), 0, GFP_KERNEL);
 	if (!skb)
 		return -ENOMEM;
@@ -1970,9 +1986,9 @@ static int ddp_setup_conn_digest(struct cxgbi_sock *csk, unsigned int tid,
 	log_debug(1 << CXGBI_DBG_TOE | 1 << CXGBI_DBG_SOCK,
 		"csk 0x%p, tid 0x%x, crc %d,%d.\n", csk, csk->tid, hcrc, dcrc);
 
-	reinit_completion(&csk->cmpl);
+	reinit_completion(cmpl);
 	cxgb4_ofld_send(csk->cdev->ports[csk->port_id], skb);
-	wait_for_completion(&csk->cmpl);
+	wait_for_completion(cmpl);
 
 	return csk->err;
 }
diff --git a/drivers/scsi/cxgbi/libcxgbi.c b/drivers/scsi/cxgbi/libcxgbi.c
index 01f5ebee3229..7e6bcdfd96ff 100644
--- a/drivers/scsi/cxgbi/libcxgbi.c
+++ b/drivers/scsi/cxgbi/libcxgbi.c
@@ -556,16 +556,21 @@ EXPORT_SYMBOL_GPL(cxgbi_sock_free_cpl_skbs);
 
 static struct cxgbi_sock *cxgbi_sock_create(struct cxgbi_device *cdev)
 {
-	struct cxgbi_sock *csk = kzalloc(sizeof(*csk), GFP_NOIO);
+	gfp_t gfp_flags = GFP_NOIO;
+	struct cxgbi_sock *csk = kzalloc(sizeof(*csk), gfp_flags);
+	struct completion *cmpl = shadow_var_alloc(csk, "shadow_cxgbi_sock", sizeof(*cmpl), gfp_flags);
 
 	if (!csk) {
 		pr_info("alloc csk %zu failed.\n", sizeof(*csk));
+		kfree(csk);
+		kfree(cmpl);
 		return NULL;
 	}
 
 	if (cdev->csk_alloc_cpls(csk) < 0) {
 		pr_info("csk 0x%p, alloc cpls failed.\n", csk);
 		kfree(csk);
+		kfree(cmpl);
 		return NULL;
 	}
 
@@ -574,7 +579,7 @@ static struct cxgbi_sock *cxgbi_sock_create(struct cxgbi_device *cdev)
 	skb_queue_head_init(&csk->receive_queue);
 	skb_queue_head_init(&csk->write_queue);
 	timer_setup(&csk->retry_timer, NULL, 0);
-	init_completion(&csk->cmpl);
+	init_completion(cmpl);
 	rwlock_init(&csk->callback_lock);
 	csk->cdev = cdev;
 	csk->flags = 0;
diff --git a/drivers/scsi/cxgbi/libcxgbi.h b/drivers/scsi/cxgbi/libcxgbi.h
index 3bf7414a75e5..2a7f99075221 100644
--- a/drivers/scsi/cxgbi/libcxgbi.h
+++ b/drivers/scsi/cxgbi/libcxgbi.h
@@ -30,6 +30,8 @@
 
 #include <libcxgb_ppm.h>
 
+#include <linux/shadow_var.h>
+
 enum cxgbi_dbg_flag {
 	CXGBI_DBG_ISCSI,
 	CXGBI_DBG_DDP,
@@ -146,7 +148,13 @@ struct cxgbi_sock {
 	struct sk_buff_head receive_queue;
 	struct sk_buff_head write_queue;
 	struct timer_list retry_timer;
-	struct completion cmpl;
+	/**
+	 * Added in 4c3b23e90307 ("scsi: cxgb4i: add
+	 * wait_for_completion()") but it doesn't fit in a hole.  We'll use
+	 * the shadow_var API.
+	 *
+	 * struct completion cmpl;
+	 */
 	int err;
 	rwlock_t callback_lock;
 	void *user_data;
@@ -296,6 +304,7 @@ static inline void cxgbi_sock_free(struct kref *kref)
 		log_debug(1 << CXGBI_DBG_SOCK,
 			"free csk 0x%p, state %u, flags 0x%lx\n",
 			csk, csk->state, csk->flags);
+		shadow_var_free(csk, "shadow_cxgbi_sock");
 		kfree(csk);
 	}
 }
@@ -487,10 +496,24 @@ struct cxgbi_device {
 	void (*csk_ddp_clear_map)(struct cxgbi_device *cdev,
 				  struct cxgbi_ppm *,
 				  struct cxgbi_task_tag_info *);
+#ifndef __GENKSYMS__
+	/**
+	 * Extra arguments removed in 4c3b23e90307 ("scsi: cxgb4i: add
+	 * wait_for_completion()") - it is fine to hide them from genksyms
+	 * as both amd64 and arm architectures, prior modules built with
+	 * the old definition and calling those functors would simply pass
+	 * one extra un-used argument in either in r8 (amd64) or x4 (arm).
+	 */
 	int (*csk_ddp_setup_digest)(struct cxgbi_sock *,
 				    unsigned int, int, int);
 	int (*csk_ddp_setup_pgidx)(struct cxgbi_sock *,
 				   unsigned int, int);
+#else
+	int (*csk_ddp_setup_digest)(struct cxgbi_sock *,
+				    unsigned int, int, int, int);
+	int (*csk_ddp_setup_pgidx)(struct cxgbi_sock *,
+				   unsigned int, int, bool);
+#endif
 
 	void (*csk_release_offload_resources)(struct cxgbi_sock *);
 	int (*csk_rx_pdu_ready)(struct cxgbi_sock *, struct sk_buff *);
-- 
2.43.0

